agent/exporter/src/main/java/com/microsoft/applicationinsights/agent/Exporter.java

```diff
@@ -116,58 +118,61 @@ private void export(SpanData span) {
             // where we capture them
             return;
         }
-        Map<AttributeKey<?>, Object> attributes = getAttributesCopy(span.getAttributes());
         if (kind == Kind.INTERNAL) {
-            boolean isLog = removeAttributeBoolean(attributes, AI_INTERNAL_LOG);
-            if (isLog) {
-                exportLogSpan(span, attributes);
+            Boolean isLog = span.getAttributes().get(AI_LOG_KEY);
+            if (isLog != null && isLog) {
+                exportLogSpan(span);
             } else if ("spring-scheduling".equals(stdComponent) && !SpanId.isValid(span.getParentSpanId())) {
                 // TODO need semantic convention for determining whether to map INTERNAL to request or dependency
                 //  (or need clarification to use SERVER for this)
-                exportRequest(span, attributes);
+                exportRequest(span);
             } else {
-                exportRemoteDependency(span, attributes, true);
+                exportRemoteDependency(span, true);
             }
         } else if (kind == Kind.CLIENT || kind == Kind.PRODUCER) {
-            exportRemoteDependency(span, attributes, false);
+            exportRemoteDependency(span, false);
         } else if (kind == Kind.CONSUMER && !span.hasRemoteParent()) {
             // TODO need spec clarification, but it seems polling for messages can be CONSUMER also
             //  in which case the span will not have a remote parent and should be treated as a dependency instead of a request
-            exportRemoteDependency(span, attributes, false);
+            exportRemoteDependency(span, false);
         } else if (kind == Kind.SERVER || kind == Kind.CONSUMER) {
-            exportRequest(span, attributes);
+            exportRequest(span);
         } else {
             throw new UnsupportedOperationException(kind.name());
         }
     }
 
-    private void exportRequest(SpanData span, Map<AttributeKey<?>, Object> attributes) {
+    private void exportRequest(SpanData span) {
 
         RequestTelemetry telemetry = new RequestTelemetry();
 
         String source = null;
-        String sourceAppId = removeAttributeString(attributes, SPAN_SOURCE_ATTRIBUTE_NAME);
+        ReadableAttributes attributes = span.getAttributes();
+        String sourceAppId = attributes.get(AI_SPAN_SOURCE_KEY);
```

Comment:
can attributes become null?

Reply:
they will be null if they were never set